tests/worker_heartbeat_test.go

```diff
@@ -0,0 +1,162 @@
+
+	tv := testvars.New(t)
+	workerInstanceKey := tv.Any().String()
+
+	resp, err := s.FrontendClient().RecordWorkerHeartbeat(ctx, &workflowservice.RecordWorkerHeartbeatRequest{
+		Namespace: s.Namespace().String(),
+		Identity:  "test-identity",
+		WorkerHeartbeat: []*workerpb.WorkerHeartbeat{
+			{
+				WorkerInstanceKey: workerInstanceKey,
+			},
+		},
+	})
+
+	require.NoError(t, err)
+	require.NotNil(t, resp)
+
+	// Verify the worker entity was created in CHASM
+	engineCtx := chasm.NewEngineContext(ctx, s.chasmEngine)
+	worker, err := chasm.ReadComponent(
+		engineCtx,
+		chasm.NewComponentRef[*chasmworker.Worker](
+			chasm.EntityKey{
+				NamespaceID: s.NamespaceID().String(),
+				BusinessID:  workerInstanceKey,
+			},
+		),
+		func(w *chasmworker.Worker, _ chasm.Context, _ any) (*chasmworker.Worker, error) {
+			return w, nil
+		},
+		nil,
+	)
+
+	require.NoError(t, err)
+	require.NotNil(t, worker)
+	require.Equal(t, workerstatepb.WORKER_STATUS_ACTIVE, worker.Status)
+	require.Equal(t, workerInstanceKey, worker.GetWorkerHeartbeat().GetWorkerInstanceKey())
+	require.NotNil(t, worker.GetLeaseExpirationTime())
+}
+
+func (s *workerHeartbeatTestSuite) TestRecordMultipleHeartbeats() {
+	t := s.T()
+	ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
+	defer cancel()
+
+	tv := testvars.New(t)
+	workerInstanceKey := tv.Any().String()
+
+	req := &workflowservice.RecordWorkerHeartbeatRequest{
+		Namespace: s.Namespace().String(),
+		Identity:  "test-identity",
+		WorkerHeartbeat: []*workerpb.WorkerHeartbeat{
+			{
+				WorkerInstanceKey: workerInstanceKey,
+			},
+		},
+	}
+
+	// Send first heartbeat
+	resp1, err := s.FrontendClient().RecordWorkerHeartbeat(ctx, req)
+	require.NoError(t, err)
+	require.NotNil(t, resp1)
+
+	// Read worker state after first heartbeat
+	engineCtx := chasm.NewEngineContext(ctx, s.chasmEngine)
+	worker1, err := chasm.ReadComponent(
+		engineCtx,
+		chasm.NewComponentRef[*chasmworker.Worker](
+			chasm.EntityKey{
+				NamespaceID: s.NamespaceID().String(),
+				BusinessID:  workerInstanceKey,
+			},
+		),
+		func(w *chasmworker.Worker, _ chasm.Context, _ any) (*chasmworker.Worker, error) {
+			return w, nil
+		},
+		nil,
+	)
+	require.NoError(t, err)
+	require.NotNil(t, worker1)
+	firstLeaseExpiration := worker1.GetLeaseExpirationTime().AsTime()
+
+	// Wait a bit to ensure lease expiration time will be different
+	time.Sleep(100 * time.Millisecond)
+
+	// Send second heartbeat for the same worker
+	resp2, err := s.FrontendClient().RecordWorkerHeartbeat(ctx, req)
+	require.NoError(t, err)
+	require.NotNil(t, resp2)
+
+	// Verify that the second heartbeat updated the existing worker entity
+	worker2, err := chasm.ReadComponent(
+		engineCtx,
+		chasm.NewComponentRef[*chasmworker.Worker](
+			chasm.EntityKey{
+				NamespaceID: s.NamespaceID().String(),
+				BusinessID:  workerInstanceKey,
+			},
+		),
+		func(w *chasmworker.Worker, _ chasm.Context, _ any) (*chasmworker.Worker, error) {
+			return w, nil
```

Comment:
Avoid the data race:

```suggestion
			return w.Clone(), nil
```